---
title: "The RVS (Rare Variant Sharing) Package"
author: "Thomas Sherman"
data: "`r Sys.Date()`"
package: "`r pkg_ver('RVsharing2')`"
bibliography: References.bib
vignette: >
    %\VignetteIndexEntry{The RVS Package}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
output:
    BiocStyle::html_document
---

```{r include=FALSE, cache=FALSE}
library(RVS)
library(kinship2)
```

# Introduction

The primary use of the *RVS* package is to compute rare variant sharing probabilities outlined in @METHODS.

# Pedigree format

The primary input used in this package is a *Pedigree* object from the *kinship2* package on CRAN (@KINSHIP_PKG). The package vignette outlines the basic steps to creating a *Pedigree* object, see here (https://cran.r-project.org/web/packages/kinship2/vignettes/pedigree.pdf). Only the *id*, *findex*, *mindex* fields are necessary for computing sharing probabilities. The *RVS* package comes with 8 sample pedigrees.

```{r}
data(samplePedigrees) # load sample pedigrees
kinship2::plot.pedigree(samplePedigrees$firstCousinPair) # plot pedigree
```

# Computing Standard Sharing Probabilities

The primary function is *RVsharing*. Most of the package features are accessible through this function. 

## Assuming One Founder Introduces the Variant

```{r}
data(samplePedigrees)
firstCousinProb <- RVsharing(samplePedigrees$firstCousinPair)
```

## When Allele Frequency is Known in the Founder Population

```{r}
defaultProbs <- c()
suppressMessages(defaultProbs[1] <- RVsharing(samplePedigrees$firstCousinPair))
suppressMessages(defaultProbs[2] <- RVsharing(samplePedigrees$secondCousinPair))
suppressMessages(defaultProbs[3] <- RVsharing(samplePedigrees$firstCousinTriple))
suppressMessages(defaultProbs[4] <- RVsharing(samplePedigrees$secondCousinTriple))

exactProbs <- list()
freq <- c(0.001,0.0025,0.005,0.01,0.025,0.05)
exactProbs$fistCousinPair <- sapply(freq, function(f) suppressMessages(RVsharing(samplePedigrees$firstCousinPair, alleleFreq=f)))
exactProbs$secondCousinPair <- sapply(freq, function(f) suppressMessages(RVsharing(samplePedigrees$secondCousinPair, alleleFreq=f)))
exactProbs$firstCousinTriple <- sapply(freq, function(f) suppressMessages(RVsharing(samplePedigrees$firstCousinTriple, alleleFreq=f)))
exactProbs$secondCousinTriple <- sapply(freq, function(f) suppressMessages(RVsharing(samplePedigrees$secondCousinTriple, alleleFreq=f)))

plot(NULL, xlim=c(0.001,0.05), ylim=c(0,0.12),log='x',xaxt='n', ylab='probability of sharing', xlab='variant frequency [%]')
axis(side=1, at=freq, labels=freq*100)
invisible(sapply(exactProbs, lines, x=freq))
invisible(sapply(exactProbs, points, x=freq, pch=19))
abline(h=defaultProbs,lty=2)
```

Similiar to Figure 2 in @METHODS

## Sharing Probabilties in a Subset of a Pedigree

# Correcting for Related Founders

When founders of the pedigree are related, the computation is more tricky. 'RVsharing' allows the user to apply a correction for this fact in two different ways.

## Exact Computation with Kinship Coefficient

In this method, a mean kinship coefficient among the founders is given. Using the methods from @METHODS, RVsharing then computes the sharing probability on the assumption one or two founders introduce the variant. Weighting each probability using a calculation based on the mean kinship coefficient. To use this correction, provide the parameter 'kinshipCoeff'.

```{r}
print(RVsharing(samplePedigrees$firstCousinPair, kinshipCoeff = 0.05))
kCoeff <- seq(0.005,0.15,0.005)
sharingProb <- sapply(kCoeff, function(c) RVsharing(samplePedigrees$firstCousinPair, kinshipCoeff=c))
plot(kCoeff, sharingProb, type='l')
```

## Estimating Kinship Coefficient Among Founders

# Using Monte Carlo Simulation

'RVsharing' also allows for estimating sharing probabilities through monte carlo simulation. The primary use of this feature is for calculating sharing probabilities under non standard assumptions about the founders. See [Calculating P-Values across multiple families](#Calculating P Values across multiple families). However, this feature is available for the standard assumptions as well. To run a monte carlo simulation, specify all parameters as normal and additionally provide the 'nSimulations' parameter.

```{r}
print(RVsharing(samplePedigrees$firstCousinPair))
print(RVsharing(samplePedigrees$firstCousinPair, nSim=1e4))
```

## Standard Sharing Probabilities

## Custom Founder Distributions 

This method allows for more complex relationships among the founders to be given. 'RVsharing' allows for a complete distribution among the founders to be passed in as a parameter - 'founderDist'. This function should accept a single argument, N, and should return a vector of length N with values in {0,1,2} representing the number of copies of the variant each founder has.

```{r}
fDist <- function(N) sample(c(rep(0,N-1), 1)) # assumption that 1 founder introduces
print(RVsharing(samplePedigrees$firstCousinPair, nSim=1e4, founderDist=fDist))
print(RVsharing(samplePedigrees$firstCousinPair))
```

# Calculating Sharing Probabilties Across Multiple Families

For variants seen in only one family, the RVsharing probability can be interpreted directly as a P-value from a Bernoulli trial. For variants seen in M families and shared by affected relatives in a subset S of them, the P-value can be obtained as the sum of the probability of events as (or more) extreme as the observed sharing in the family subset S. 
```{r}
variantSharedPeds <- list(samplePedigrees$firstCousinPair, samplePedigrees$secondCousinPair)
variantNotSharedPeds <- list(samplePedigrees$firstCousinTriple)
sharedProbs <- sapply(variantSharedPeds, RVsharing)
notSharedProbs <- sapply(variantNotSharedPeds, RVsharing)
#multipleFamilyPValue(sharedProbs, notSharedProbs)
```

# Sharing Probabilities in a Genomic Region

# References
